workflow/rules/common.smk

```diff
@@ -28,27 +29,128 @@ validate(samples, schema="../schemas/samples.schema.yaml")
 
 ### Read and validate units file
 
-units = pd.read_table(config["units"], dtype=str).set_index(
-    ["sample", "unit", "run", "lane"], drop=False
-)
+units = pd.read_table(config["units"], dtype=str).set_index(["sample", "unit", "run", "lane"], drop=False)
 validate(units, schema="../schemas/units.schema.yaml")
 
 ### Set wildcard constraints
 
 
 wildcard_constraints:
     sample="|".join(samples.index),
-
-def get_fastq(wildcards):
-    fastqs = units.loc[
-        (wildcards.sample, wildcards.unit, wildcards.run, wildcards.lane), ["fastq1", "fastq2"]
-    ].dropna()
-    return {"fwd": fastqs.fastq1, "rev": fastqs.fastq2}
-
-
-def get_sample_fastq(wildcards):
-    fastqs = units.loc[(wildcards.sample, wildcards.unit), ["fastq1", "fastq2"]].dropna()
-    return {"fwd": fastqs["fastq1"].tolist(), "rev": fastqs["fastq2"].tolist()}
-
-def compile_output_list(wildcards):
-    return ["dummy.txt"]
+    unit="N|T|R",
+    read="fastq[1|2]",
+
+
+def get_unit(units: pd.DataFrame, wildcards: snakemake.io.Wildcards) -> pd.Series:
+    """
+    function used to extract one unit(row) from units.tsv
+
+    Args:
+        units: DataFrame generate by importing a file following schema defintion
+               found in pre-alignment/workflow/schemas/units.schema.tsv
+        wildcards: wildcards object with at least the following widlcard names
+               sample, unit, run, and lane
+
+    Returns:
+        Series containing data of the selected row
+
+    Raises:
+        raises and exception if no unit can be extracted from the Dataframe
+    """
+    unit = units.loc[(wildcards.sample, wildcards.unit, wildcards.run, wildcards.lane)].dropna()
+    return unit
+
+
+def get_fastq_file(units: pd.DataFrame, wildcards: snakemake.io.Wildcards, read_pair: str = "fastq1") -> str:
+    """
+    function used to extract path for one unit(row) from units.tsv
+
+    Args:
+        units: DataFrame generate by importing a file following schema defintion
+               found in pre-alignment/workflow/schemas/units.schema.tsv
+        wildcards: wildcards object with at least the following widlcard names
+               sample, unit, run, and lane
+        read_pair: fast1 or fastq2
+
+    Returns:
+        path for fastq file as a str
+
+    Raises:
+        raises and exception if no unit can be extracted from the Dataframe
+    """
+    unit = get_unit(units, wildcards)
+    if len(unit) == 0:
+        raise ValueError("No data found!")
+    return unit[read_pair]
+
+
+def get_fastq_adapter(units: pd.DataFrame, wildcards: snakemake.io.Wildcards) -> str:
```

Comment:
This function and the function above are doing rather similar things. I would prefer a drier version and call it something like `get_fastq_column` or similar and the string `"adapter"` can be the input at the appropriate place